File: Fwd_Euler/driver_fwd_euler.cpp
/* Main routine to test the forward Euler method for some scalar-valued ODE problems
     y' = f(t,y), t in [0,5],
     y(0) = y0.

   D.R. Reynolds
   Math 6321 @ SMU
   Fall 2020 */

#include <iostream>
#include "fwd_euler.hpp"

using namespace std;
using namespace arma;


// Problem 1:  y' = -y
//    ODE RHS function class -- instantiates a RHSFunction
class RHS1: public RHSFunction {
public:
  int Evaluate(double t, vec& y, vec& f) {    // evaluates the RHS function, f(t,y)
    f = -y;
    return 0;
  }
};
//    Convenience function for analytical solution
vec ytrue1(const double t) {
  vec yt(1);
  yt(0) = exp(-t);
  return yt;
};



// Problem 2:  y' = (y+t^2-2)/(t+1)
//    ODE RHS function class -- instantiates a RHSFunction
class RHS2: public RHSFunction {
public:
  int Evaluate(double t, vec& y, vec& f) {    // evaluates the RHS function, f(t,y)
    f(0) = (y(0)+t*t-2.0)/(t+1.0);
    return 0;
  }
};
//    Convenience function for analytical solution
vec ytrue2(const double t) {
  vec yt(1);
  yt(0) = t*t + 2.0*t + 2.0 - 2.0*(t+1.0)*log(t+1.0);
  return yt;
};



// main routine
int main() {

  // time steps to try
  vec h("1.0 0.1 0.01 0.001 0.0001");

  // set problem information
  vec y0_1("1.0");
  vec y0_2("2.0");
  double t0 = 0.0;
  double Tf = 5.0;

  // set desired output times
  int Nout = 6;  // includes initial condition
  vec tspan = linspace(t0, Tf, Nout);

  // create ODE RHS function objects
  RHS1 f1;
  RHS2 f2;

  // create true solution results
  mat Y1true(1,Nout), Y2true(1,Nout);
  for (size_t i=0; i<Nout; i++) {
    Y1true.col(i) = ytrue1(tspan(i));
    Y2true.col(i) = ytrue2(tspan(i));
  }

  // problem 1: loop over time step sizes; call stepper and compute errors
  cout << "\nProblem 1:\n";
  ForwardEulerStepper FE1(f1, y0_1);
  for (size_t ih=0; ih<h.n_elem; ih++) {

    // call stepper
    cout << "  h = " << h(ih) << ":\n";
    mat Y = FE1.Evolve(tspan, h(ih), y0_1);

    // output solution, errors, and overall error
    mat Yerr = abs(Y-Y1true);
    for (size_t i=0; i<Nout; i++)
      printf("    y(%.1f) = %9.6f   |error| = %.2e\n",
             tspan(i), Y(0,i), Yerr(0,i));
    cout << "  Max error = " << Yerr.max() << endl;

  }

  // problem 2: loop over time step sizes; call stepper and compute errors
  cout << "\nProblem 2:\n";
  ForwardEulerStepper FE2(f2, y0_2);
  for (size_t ih=0; ih<h.n_elem; ih++) {

    // call stepper; output solution and error
    cout << "  h = " << h(ih) << ":\n";
    mat Y = FE2.Evolve(tspan, h(ih), y0_2);

    // output solution, errors, and overall error
    mat Yerr = abs(Y-Y2true);
    for (size_t i=0; i<Nout; i++)
      printf("    y(%.1f) = %9.6f   |error| = %.2e\n",
             tspan(i), Y(0,i), Yerr(0,i));
    cout << "  Max error = " << Yerr.max() << endl;

  }

  return 0;
}

File: forward_euler/driver_fwd_euler.py
#!/usr/bin/env python3
#
# Main routine to test the forward Euler method on two scalar-valued ODE problems
#    y' = -y, t in [0,5],
#    y(0) = 1.
# and
#    y' = (y+t^2-2)/(t+1), t in [0,5],
#    y(0) = 2.
#
# D.R. Reynolds
# Math 6321 @ SMU
# Fall 2023

import numpy as np
from ForwardEuler import *

# problem time interval
t0 = 0.0
tf = 5.0

# problem-definining functions and initial conditions
def f1(t,y):
    """ ODE RHS function """
    return -y
def ytrue1(t):
    """ Analytical solution """
    return np.array([np.exp(-t)])

def f2(t,y, alpha, beta):
    """ ODE RHS function (with parameters alpha and beta)"""
    return np.array([(alpha*y[0]+t*t-2.0*beta)/(t+1.0)])
def ytrue2(t):
    """ Analytical solution """
    return np.array([t*t + 2.0*t + 2.0 - 2.0*(t+1.0)*np.log(t+1.0)])

# shared testing data
Nout = 6   # includes initial condition
tspan = np.linspace(t0, tf, Nout)

# create true solution results
Y1true = np.zeros((Nout,1))
Y2true = np.zeros((Nout,1))
for i in range(Nout):
    Y1true[i,:] = ytrue1(tspan[i])
    Y2true[i,:] = ytrue2(tspan[i])

# time steps to try
hvals = np.array([0.5, 0.05, 0.005, 0.0005, 0.00005])
errs = np.zeros(hvals.size)

# problem 1: loop over time step sizes; call stepper and compute errors
print("\nProblem 1:")
FE1 = ForwardEuler(f1)
for idx, h in enumerate(hvals):

    # set initial condition and call stepper
    y0 = Y1true[0,:]
    print("  h = ", h, ":")
    FE1.reset()
    Y, success = FE1.Evolve(tspan, y0, h)

    # output solution, errors, and overall error
    Yerr = np.abs(Y-Y1true)
    errs[idx] = np.linalg.norm(Yerr,np.inf)
    for i in range(Nout):
        print("    y(%.1f) = %9.6f   |error| = %.2e" % (tspan[i], Y[i,0], Yerr[i,0]))
    print("  overall:  steps = %5i  abserr = %9.2e\n" % (FE1.get_num_steps(), errs[idx]))
orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
print('estimated order: max = ', np.max(orders), ',  avg = ', np.average(orders))


# problem 2: loop over time step sizes; call stepper and compute errors
print("\nProblem 2:")
FE2 = ForwardEuler(f2)
for idx, h in enumerate(hvals):

    # set initial condition and call stepper
    y0 = Y2true[0,:]
    print("  h = ", h, ":")
    FE2.reset()
    alpha = 1.0
    beta = 1.0
    # Here when calling our rhs (f2) notice we have parameters alpha
    # and beta. At within the ForwardEuler.py we will call f2(t,y,*args)
    # Here *(<tuple>) is the unpacking operator. This will unpack the args tuple into
    # the third and fourth arguments of f2. This is the same implementation
    # as in standard python ODE packages.
    Y, success = FE2.Evolve(tspan, y0, h, args=(alpha, beta))

    # output solution, errors, and overall error
    Yerr = np.abs(Y-Y2true)
    errs[idx] = np.linalg.norm(Yerr,np.inf)
    for i in range(Nout):
        print("    y(%.1f) = %9.6f   |error| = %.2e" % (tspan[i], Y[i,0], Yerr[i,0]))
    print("  overall:  steps = %5i  abserr = %9.2e  relerr = %9.2e\n" %
          (FE2.get_num_steps(), errs[idx], np.linalg.norm(Yerr/Y2true,np.inf)))
orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
print('estimated order: max = ', np.max(orders), ',  avg = ', np.average(orders))
